File: 2021/rust/src/hydrothermalventure.rs
use std::collections::HashMap;

type V2 = (i32, i32);

fn parse(line: &str) -> (V2, V2) {

    let parts: Vec<&str> = line.split(" -> ").collect();

    let start: Vec<&str> = parts[0].split(",").collect();
    let end: Vec<&str> = parts[1].split(",").collect();

    (
        (start[0].parse::<i32>().unwrap(),
         start[1].parse::<i32>().unwrap()),
        (end[0].parse::<i32>().unwrap(),
         end[1].parse::<i32>().unwrap())
    )
}

fn _print_grid(grid: HashMap<V2, u32>) {

    for row in 0..10 {
        for col in 0..10 {
            let val = grid.get(&(row,col));
            match val {
                Some(v) => {
                    if *v == 0 {
                        print!(".");
                    } else {
                        print!("{}", grid[&(row,col)]);
                    }
                },
                None => { print!(".") }
            };
        }
        println!();
    }
}


fn mark_points(points: &Vec<(V2, V2)>, skip_diagonals: bool) -> u32 {

    let mut grid: HashMap<V2, u32> = HashMap::new();

    for (x, y) in points {
        let (x1, y1) = x;
        let (x2, y2) = y;

        if skip_diagonals && !(x1 == x2 || y1 == y2) {
            continue;
        }

        let dx = x2 - x1;
        let dy = y2 - y1;
        let length = dx.abs().max(dy.abs());
        let step_x = dx / length;
        let step_y = dy / length;
        for i in 0..(length + 1) {
            let v = grid.entry((x1 + i * step_x, y1 + i * step_y)).or_insert(0);
            *v += 1;
        }
    }

    grid.values().filter(|&v| v > &1).collect::<Vec<&u32>>().len() as u32

}

// Day 5: Hydrothermal Venture
pub fn solve(data: String) {

    let points: Vec<(V2, V2)> = data.split("\n").map(|l| parse(l)).collect();

    println!("{}", mark_points(&points, true));
    println!("{}", mark_points(&points, false));

}


#[cfg(test)]
mod tests {

    use super::*;
    use std::fs;

    fn load_input_file(filename: &str) -> String {
        fs::read_to_string(filename).unwrap().trim().to_string()
    }

    #[test]
    fn testinput() {


        let points = vec![((0, 9), (5, 9)),
                          ((8, 0), (0, 8)),
                          ((9, 4), (3, 4)),
                          ((2, 2), (2, 1)),
                          ((7, 0), (7, 4)),
                          ((6, 4), (2, 0)),
                          ((0, 9), (2, 9)),
                          ((3, 4), (1, 4)),
                          ((0, 0), (8, 8)),
                          ((5, 5), (8, 2))];

        assert_eq!(mark_points(&points, true), 5);
        assert_eq!(mark_points(&points, false), 12);
    }

    #[test]
    fn solution() {

        let input = load_input_file("../inputs/input05.txt");
        let points: Vec<(V2, V2)> = input.split("\n").map(|l| parse(l)).collect();

        assert_eq!(mark_points(&points, true), 5306);
        assert_eq!(mark_points(&points, false), 17787);
    }
}

File: 2021/haskell/src/HydrothermalVenture.hs
module HydrothermalVenture where

import qualified Data.Map        as M
import           Data.List.Split (splitOn)

type V2 = (Int, Int)

parse :: String -> (V2, V2)
parse line = ((x1, y1), (x2, y2))
  where (a:_:b:[]) = words line
        (x1:y1:[]) = (map read) $ splitOn "," a
        (x2:y2:[]) = (map read) $ splitOn "," b

generatePoints :: (V2, V2) -> [V2]
generatePoints ((x1, y1), (x2, y2)) = map nextPoint [0..length]
  where nextPoint d = (x1 + d * stepX, y1 + d * stepY)
        stepX  = dx `div` length
        stepY  = dy `div` length
        length = max (abs dx) (abs dy)
        dx     = x2 - x1
        dy     = y2 - y1

markPoint :: M.Map V2 Int -> V2 -> M.Map V2 Int
markPoint grid point = case M.lookup point grid of
  Nothing -> M.insert point 1 grid
  Just v  -> M.adjust succ point grid

solve :: [(V2, V2)] -> Int
solve lines = M.size
              $ M.filter (> 1)
              $ foldl markPoint M.empty (concatMap generatePoints lines)

solve1 :: [(V2, V2)] -> Int
solve1 = solve . filter (\((x1, y1),(x2,y2)) -> x1 == x2 || y1 == y2)

day05:: [String] -> IO ()
day05 input = do
  let lines = map parse input
  print $ solve1 lines
  print $ solve lines
